test/jdk/java/net/Socket/CtorFailTest.java

```diff
@@ -0,0 +1,215 @@
+import java.net.SocketImpl;
+import java.net.SocketImplFactory;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+/*
+ * @test
+ * @summary verifies that the socket is closed on constructor failures
+ * @library /test/lib
+ * @run junit/othervm --add-opens java.base/java.net=ALL-UNNAMED CtorFailTest
+ */
+class CtorFailTest {
+
+    private static final VarHandle SOCKET_IMPL_FACTORY_HANDLE = createSocketImplFactoryHandle();
+
+    private static final int DEAD_SERVER_PORT = 0xDEAD;
+
+    private static VarHandle createSocketImplFactoryHandle() {
+        try {
+            Field field = Socket.class.getDeclaredField("factory");
+            field.setAccessible(true);
+            return MethodHandles
+                    .privateLookupIn(Socket.class, MethodHandles.lookup())
+                    .findStaticVarHandle(Socket.class, "factory", SocketImplFactory.class);
+        } catch (NoSuchFieldException | IllegalAccessException exception) {
+            throw new RuntimeException(exception);
+        }
+    }
+
+    private static void withSocketImplFactory(SocketImplFactory newFactory, Executable executable) throws Throwable {
+        SocketImplFactory oldFactory = (SocketImplFactory) SOCKET_IMPL_FACTORY_HANDLE.getAndSet(newFactory);
+        try {
+            executable.execute();
+        } finally {
+            SOCKET_IMPL_FACTORY_HANDLE.set(oldFactory);
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("testCases")
+    @SuppressWarnings("resource")
+    void test(Exception bindException, Exception connectException) throws Throwable {
+
+        // Create a socket using the mock `SocketImpl` configured to fail
+        MockSocketImpl socketImpl = new MockSocketImpl(bindException, connectException);
+        withSocketImplFactory(() -> socketImpl, () -> {
+
+            // Trigger the failure
+            Exception caughtException = assertThrows(Exception.class, () -> {
+                // Address and port are mostly ineffective.
+                // They just need to be _valid enough_ to reach to the point where both `SocketImpl#bind()` and `SocketImpl#connect()` are invoked.
+                // Failure will be triggered by the injected `SocketImpl`.
+                InetAddress serverAddress = InetAddress.getLoopbackAddress();
+                new Socket(serverAddress, DEAD_SERVER_PORT, null, 0);
+            });
+
+            // Run verifications
+            Exception expectedException = bindException != null ? bindException : connectException;
+            assertSame(expectedException, caughtException);
+            assertEquals(1, socketImpl.closeInvocationCounter.get());
+
+        });
+
+    }
+
+    static List<Arguments> testCases() {
+        String exceptionMessage = "intentional test failure";
```

Comment:
It would be good to put a comment here to remind the reader what the first argument is and what the second argument is. Something like:

```
        // Arguments: Exception bindException, Exception connectException
```

Reply:
Switched from using (untyped) `Arguments` to `MockSocketImpl` in 5e492b7bb83c6668f31d3e72aa6ce458221ee846.